File: microservices/enrollment-service/src/main/java/com/karate/enrollment_service/infrastructure/jwt/SecurityConfig.java
package com.karate.enrollment_service.infrastructure.jwt;

import lombok.AllArgsConstructor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.Customizer;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;
import org.springframework.web.cors.UrlBasedCorsConfigurationSource;

import java.util.List;

@Configuration
@AllArgsConstructor
public class SecurityConfig {
    private final JwtAuthTokenFilter jwtAuthTokenFilter;

    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity httpSecurity) throws Exception {
        return httpSecurity
                .cors(AbstractHttpConfigurer::disable)
                .csrf(AbstractHttpConfigurer::disable)
                .authorizeHttpRequests(auth -> auth
                        .requestMatchers("/").permitAll()
                        .requestMatchers("/swagger-ui/**").permitAll()
                        .requestMatchers("/swagger-ui.html/**").permitAll()
                        .requestMatchers("/v3/api-docs*/**").permitAll()
                        .requestMatchers("/v2/api-docs*/**").permitAll()
                        .requestMatchers("/webjars/**").permitAll()
                        .requestMatchers("/swagger-resources/**").permitAll()
                        .requestMatchers("/internal/enrollments/**").permitAll()
                        .requestMatchers("/enrollments/me").hasAnyRole("USER", "ADMIN")
                        .requestMatchers("/enrollments/me/**").hasAnyRole("USER", "ADMIN")
                        .requestMatchers("/enrollments/training/{trainingId}").hasRole("ADMIN")
                        .requestMatchers("/enrollments/{userId}/{trainingId}").hasRole("ADMIN")
                        .requestMatchers("/enrollments/user/{userId}").hasRole("ADMIN")
                        .anyRequest().authenticated()
                )
                .sessionManagement(sess -> sess.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
                .httpBasic(Customizer.withDefaults())
                .addFilterBefore(jwtAuthTokenFilter, UsernamePasswordAuthenticationFilter.class)
                .build();
    }
}

File: microservices/feedback-service/src/main/kotlin/com/karate/feedback_service/infrastructure/jwt/SecurityConfig.kt
package com.karate.feedback_service.infrastructure.jwt;

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Profile
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter

@Profile("!test")
@Configuration
class SecurityConfig(
    private val jwtAuthTokenFilter: JwtAuthTokenFilter
) {

    @Bean
    fun securityFilterChain(http: HttpSecurity): SecurityFilterChain =
        http
            .cors { it.disable() }
            .csrf { it.disable() }
            .authorizeHttpRequests {
                it.requestMatchers(
                    "/", "/swagger-ui/**", "/swagger-ui.html/**",
                    "/v3/api-docs*/**", "/v2/api-docs*/**",
                    "/webjars/**", "/swagger-resources/**"
                ).permitAll()
                    .requestMatchers("/feedbacks/*").hasAnyRole("USER", "ADMIN")
                    .requestMatchers("/feedbacks/me").hasAnyRole("USER","ADMIN")
                    .requestMatchers("/feedbacks/admin/**").hasRole("ADMIN")
                    .requestMatchers("/feedbacks/*/*").hasRole("ADMIN")
                    .anyRequest().authenticated()
            }
            .sessionManagement { it.sessionCreationPolicy(SessionCreationPolicy.STATELESS) }
            .httpBasic {}
            .addFilterBefore(jwtAuthTokenFilter, UsernamePasswordAuthenticationFilter::class.java)
            .build()
}
